dvc/stage.py

```diff
@@ -395,11 +402,14 @@ def reproduce(self, interactive=False, **kwargs):
 
         return self
 
-    def update(self):
+    def update(self, rev=None):
         if not self.is_repo_import and not self.is_import:
             raise StageUpdateError(self.relpath)
 
-        self.deps[0].update()
+        if not self.is_repo_import and rev:
+            raise UpdateWithRevNotPossibleError()
+
```

Comment:
We  could move it even more deep to get rid of is_repo_import check. E.g. we could move
```
if rev:
    raise ...
```
to OutputBASE.update() method, since it is the one overwritten by DependencyREPO.

But that is me jumping into implementation details without asking a more high-level questions first :slightly_smiling_face:  Here is one: maybe we should consider simply ignoring `rev` for non-repo imports? E.g. `dvc update repo-import.dvc non-repo-import.dvc` is a valid command and currently it updates both fine, because `Output.update()` for the latter one does nothing. What do you think?

Reply:
Thanks Reslan!

I think it is better to stay explicit instead of silently ignoring invalid uses including for `dvc update repo-import.dvc non-repo-import.dvc`.

1- without an error, user will assume that command actually did what (s)he intent to do but the result will be different and this is an ambiguous behavior.

2- by raising an error, we prevent 1- to happen and teach user the correct usage of the command.

Reply:
Btw, I cannot see an `update` method under [OutputBASE](https://github.com/iterative/dvc/blob/master/dvc/output/base.py), maybe you did mean creating one?

Sorry, I don't know much about DVC code yet to give some opinionated ideas.

But I think `Stage.update()` still a good place to do this check since `Stage` is a kind of box(wrapper) that deals with all different data(repo or url) sources

 It makes sense for me to do the check at this level, and run other low level `update` safely without a need of validation with the thinking of low level funcs should always receive valid parameters.

I might be wrong for this case considering my current know-how on the code base! 👨‍🚀

Reply:
> I think it is better to stay explicit instead of silently ignoring invalid uses including for dvc update repo-import.dvc non-repo-import.dvc.

That is reasonable. Let's keep the behaviour that way then.

> Btw, I cannot see an update method under OutputBASE, maybe you did mean creating one?

Sorry, I meant `DependencyBASE.update` 

https://github.com/iterative/dvc/blob/master/dvc/dependency/base.py#L30 .

> But I think Stage.update() still a good place to do this check since Stage is a kind of box(wrapper) that deals with all different data(repo or url) sources
> It makes sense for me to do the check at this level, and run other low level update safely without a need of validation with the thinking of low level funcs should always receive valid parameters.

That is fair, but if we do move that `raise` to DependencyBASE.update we won't need to check if this is a repo import, which is 1 less check. Plus, if anyone will try to use `DependencyLOCAL/s3/GS/etc.update()` he will get slapped by that exception, which is nice as it will prevent incorrect usage. So seems like a more optimal(marginally but still) and safer approach. :slightly_smiling_face:

Reply:
thanks! updated by 1562ca4df5458fdcdecba099cba35ed371bc012d